Add regex patterns to JSON schema for `Decimal` type

Dima-Bulavenko opened the pull request:
<!-- Thank you for your contribution! -->
<!-- Unless your change is trivial, please create an issue to discuss the change before creating a PR -->

## Change Summary

This PR adds a regular expression pattern to the JSON schema of `Decimal` fields to validate their string representation according to the `max_digits` and `decimal_places` parameters of the `Field` class.

- Updated existing tests for `Decimal` fields to align with the new validation pattern.
- Added new tests to validate the pattern across different combinations of `max_digits` and `decimal_places`.

You can run the new tests with the following command:  
```bash
uv run pytest tests/test_json_schema.py::TestDecimalPattern
```

## Related issue number
fixes #11016
fixes #11420
fixes #10867
<!-- WARNING: please use "fix #123" style references so the issue is closed when this PR is merged. -->

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**


Selected Reviewer: @Viicos

Dima-Bulavenko committed:
feat: Add regular expression constraint to decimal JSON schema

Dima-Bulavenko committed:
test: update existing tests for Decimal JSON schema pattern validation

Old tests have been updated to align with the new validation pattern
for Decimal fields in the JSON schema.

Dima-Bulavenko committed:
test: Add ADDTESTS comment to mark old test that have to be updated with new test cases.

Dima-Bulavenko committed:
test: Tailor print in docs for Model.model_json_schema() so that it contain validaton pattern.

Dima-Bulavenko committed:
test: Create tests to check the validation pattern for decimal string.

Dima-Bulavenko committed:
fix: Remove type annotations to pass test on Python version 3.9

Viicos reviewed (commented):
Thanks for the contribution. I've checked locally and the regex seems to correctly match.

Similar to https://github.com/pydantic/pydantic/pull/11420, we'd have to also add patterns when `max_digits` xor `decimal_places` is None. Would you be interested in pursuing?

Viicos commented on tests/test_json_schema.py:
```diff
@@ -6811,3 +6946,147 @@ def test_json_schema_arguments_v3_aliases() -> None:
     }
+
+
+class TestDecimalPattern:
```
Let's flatten this to use functions instead of methods.

Viicos commented on pydantic/json_schema.py:
```diff
@@ -674,7 +674,31 @@ def decimal_schema(self, schema: core_schema.DecimalSchema) -> JsonSchemaValue:
+                    integer_places = 0
+
+            pattern = (
+                r'^(?!^[+-\.]*$)'  # check string is not empty and not single or sequence of ".+-" characters.
```
I think you don't need to escape the `.` when inside brackets.

Dima-Bulavenko committed:
refactor: Delete unwanted escaping "\" character from regex.

refactor: Update tests to match the new regex

Dima-Bulavenko committed:
refactor: Flatten tests

Viicos commented:
> It seems the pattern already handles cases where `max_digits` or `decimal_places` is `None`.

Sorry what I said was confusing, I wanted to say that it would be nice to have simpler patterns generated for cases where `max_digits` xor `decimal_places` is `None`.

Dima-Bulavenko committed:
refactor: Reduce size of regex.

Dima-Bulavenko committed:
refactor: Tailor old test to match reduced patterns
refactor: Delete redundunt tests.
test: Create new tests

Dima-Bulavenko committed:
fix: Fix decimal pattern validation when max_digits and decimal_places are both set

The pattern incorrectly counted trailing zeros in integer strings towards max_digits.
For example, "1200" with max_digits=4 and decimal_places=2 would incorrectly match even thought pattern must allow only two integer places.

Dima-Bulavenko commented:
> > It seems the pattern already handles cases where `max_digits` or `decimal_places` is `None`.
> 
> Sorry what I said was confusing, I wanted to say that it would be nice to have simpler patterns generated for cases where `max_digits` xor `decimal_places` is `None`.

Thank you for clarifying your request.

I've updated the implementation to generate simpler patterns for each combination of **max_digits** and **decimal_places**.

Viicos commented on pydantic/json_schema.py:
```diff
@@ -674,7 +674,49 @@ def decimal_schema(self, schema: core_schema.DecimalSchema) -> JsonSchemaValue:
+
+            # Case 4: Both are None (no restrictions)
+            else:
+                pattern += r'\d*\.?\d*$'  # look for arbitrary integer or decimal
```
This allows `'.'` as an input, which can't be validated.

Viicos commented on pydantic/json_schema.py:
```diff
@@ -674,7 +674,49 @@ def decimal_schema(self, schema: core_schema.DecimalSchema) -> JsonSchemaValue:
+                )
+
+            # Case 2: Only max_digits is set
+            elif max_digits is not None and decimal_places is None:
```
With `max_digits` set to e.g. 5, it wrongfully matches `'1000.1111111'`, etc.

Dima-Bulavenko replied:
Thank you for your feedback.

From my testing, the pattern appears to work correctly when `max_digits=5` and the invalid value `'1000.1111111'` is rejected as expected.

Here is a test demonstrating this behavior:
```python
@pytest.fixture
def get_decimal_pattern():
    def pattern(max_digits=None, decimal_places=None) -> str:
        field = TypeAdapter(Annotated[Decimal, Field(max_digits=max_digits, decimal_places=decimal_places)])
        return field.json_schema()['anyOf'][1]['pattern']
    return pattern

@pytest.mark.parametrize('invalid_decimal', ['1000.1111111'])
def test_only_max_digits_set(invalid_decimal, get_decimal_pattern):
    pattern = get_decimal_pattern(max_digits=5, decimal_places=None)
    assert re.fullmatch(pattern, invalid_decimal) is None
```

Let me know if there's a specific case I may have missed!

Viicos reviewed (commented):
Given the complexity to implement this correctly (see my comments above), I'm starting to think if we should just define a simpler regex, describing the allowed characters (something like `^(?:\+|-)?[\d.]+$` in validation mode, and `^-?[\d.]+$` in serialization mode). This regex will also allow invalid inputs (but at least we are certain valid inputs wont be disallowed).

Dima-Bulavenko commented:
I think in validation mode, it's better to use the pattern `^[-+]?(?:\d+|\d+\.\d+)$`, which allows both integers and decimals of any size. However, if we use this simple pattern, we completely lose the main purpose of this issue (#10867), which is to check for `max_digits` and `decimal_places`.

Also, the more complex regex patterns are only created in two cases:
1. When `max_digits` is set, or
2. When both `max_digits` and `decimal_places` are not `None`.

I believe these cases don't come up very often, but if a user sets these options, they probably want the regex pattern to actually check `max_digits` and `decimal_places`.

For serialization mode, I agree that a simple regex like `^-?(?:\d+|\d+\.\d+)$` is useful because it shows that the data can be either an integer or a decimal.

Please let me know what you think about this. I should also mention that it's possible I don't fully understand the purpose of the validation and serialization schemas or all the cases where they might be used.

Viicos committed:
Update test_json_schema.py

Viicos reviewed (approved):
Thanks @Dima-Bulavenko, @Jack70248 I'm adding you as a co-author for the initial work.
